coalib/parsing/filters/decorators.py

```diff
@@ -0,0 +1,42 @@
+from inspect import getmro
+from functools import wraps
+
+
+def typed_filter(type_classes, msg=None):
+    """
+    Used to enforce type of the first argument of a method
+    by its class name as string. Raises NotImplementedError
+    if the type does not match. Primarily used to enforce
+    type of a filter.
+
+    :param type_classes:    list or tuple of acceptable class names
+                            as strings for the first argument of the
+                            method being decorated.
+    :param msg:             Message to pass with NotImplementedError.
+    """
+    if (type(type_classes) not in (tuple, list)):
+        type_classes = (type_classes,)
+
+    def decorator(filter):
+
+        @wraps(filter)
+        def decorated_filter(thing, *args, **kargs):
+            thing_class = thing.__class__
+            all_bases = list(map(lambda klass: klass.__name__,
+                                 getmro(thing_class)))
+
+            for type_class in type_classes:
+                if str(type_class) in all_bases:
+                    break
+            else:
+                raise NotImplementedError(
+                    msg or '\'{filter}\' filter can only handle {type_name}. '
```

Comment:
it was going so well until you mentioned filter here.

the thing you are wrapping has a name, which you are using here.

This decorator is a developer helper ... to catch when they have done the wrong thing.  it shouldnt give pretty names ... the caller needs to detect this and make it user-friendly.

Why should a `NotImplementedError` say "language_filter filter" .. ?  `language_filter` is better, and the backtrace is even better, so no need to tidy it up.

To help, create a subclass of `NotImplementedError` and stuff into it extra attributes to record what went wrong, such as the callee, expected classes, and the incorrect object which was encountered.   Then whoever is catching it can understand what went wrong and create a user-friendly message.